server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -82,27 +110,78 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
         for (IndexRoutingTable indexShardRouting : state.routingTable()) {
             for (int i = 0; i < indexShardRouting.size(); i++) {
                 IndexShardRoutingTable shardRouting = indexShardRouting.shard(i);
-                status.addPrimary(shardRouting.primaryShard(), shutdown);
+                status.addPrimary(shardRouting.primaryShard(), state, shutdown, includeDetails);
                 for (ShardRouting replicaShard : shardRouting.replicaShards()) {
-                    status.addReplica(replicaShard, shutdown);
+                    status.addReplica(replicaShard, state, shutdown, includeDetails);
                 }
             }
         }
 
-        return createIndicator(status.getStatus(), status.getSummary(), status.getDetails(includeDetails), status.getImpacts());
+        return new HealthIndicatorResult(
+            name(),
+            component(),
+            status.getStatus(),
+            status.getSummary(),
+            status.getDetails(includeDetails),
+            status.getImpacts(),
+            status.getUserActions(includeDetails)
+        );
     }
 
-    private static class ShardAllocationCounts {
+    // TODO: #85572 Fill in help URLs once they are finalized
+    public static final UserAction.Definition ACTION_RESTORE_FROM_SNAPSHOT = new UserAction.Definition(
+        "restore_from_snapshot",
+        Explanations.Allocation.NO_COPIES,
```

Comment:
For ease of review on Github this explanation text is as follows:

> Elasticsearch can't allocate this shard because there are no copies of its data in the cluster. Elasticsearch will allocate this shard when a node holding a good copy of its data joins the cluster. If no such node is available, restore this index from a recent snapshot.